### PR TITLE
Sync calico package updates to the ClusterResourceSet

### PR DESCRIPTION
**Pull request by 12345lcr**
### What this PR does / why we need it
Update the calico cni in `/pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico ` to fetch the newest changes form tce by running vendir sync.

Changes include:
* Dual-stack support
* Disable label scoping rule for calico-kube-controllers
* Make MTU configurable in calico
### Which issue(s) this PR fixes

Fixes: https://github.com/vmware-tanzu/tanzu-framework/issues/1275

### Release note

```release-note
None
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

#### Special notes for your reviewer

After running `vendir sync`, I excluded the updates that not for calico

### CODE REVIEW

**Comment by 12345lcr** [comment]
Wait until these changes merged into bom file and I will test again

File: pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml
Comment (vijaykatam) [inline pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: `always-replace` means kapp will delete the deployment and then install instead of an update. What is the reasoning for doing `always-replace` for a critical component like CNI?

Comment (12345lcr) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: Because after we add this `disable-default-label-scoping-rules` annotation, kapp no longer add its `labelSelector` to calico-kube-controllers. But for a cluster update with old version of calico installed who do not have this annotation, means the kapp `labelSelector` already exists. Since this field is immutable, it will leads to an update failure due to not able to remove kapp `labelSelector`.

Calico-kube-controllers is a set of controllers who only have 1 replica, which means only 1 pod and will be restarted during update anyway, and it does not obtain any volume also, so the best and only way to prevent update failure is to replace it during updating.

Comment (vijaykatam) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: If I am reading https://carvel.dev/kapp/docs/latest/faq/#migrating-from-kubectl-apply-to-kapp correctly, just the annotation should be enough to tell kapp to not touch the labels. Looping in @cppforlife for input.

Also consider using `fallback-on-replace` instead of `always-replace`

Comment (12345lcr) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: I was also considered `fallback-on-replace`, based on [kapp doc](https://carvel.dev/kapp/docs/latest/apply/#kappk14sioupdate-strategy), this fallback is used for service rolling updates who have multiple pods to prevent downtime, or pods who obtain volumes to prevent data loss, which none of them meets calico-kube-controller.

And when doing an update to an deployment, some field change(for example label or container environment variable) will not lead to a pod restart or recreate, I was worried that it may leads to any unexpected behavior in the future if we have any manifest changes but do not replace it.

Comment (vijaykatam) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: > And when doing an update to an deployment, some field change

Field changes always trigger an update by kapp, if there is an actual issue lets get that logged to kapp and get it triaged.

Going back to original issue, does just adding `disable-default-label-scoping-rules` not fix the issue you are seeing?

Comment (12345lcr) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: @vijaykatam No, we firstly added `disable-default-label-scoping-rules`, it will solve the **management cluster install** error when pkgi adopt ClusterResourceSet calico componentes. But after this annotation is added, latter we found all the  **workload cluster upgrading** is breaking with the error below.
```
[2021-11-30T03:19:47.656Z] kapp: Error: Applying update deployment/calico-kube-controllers (apps/v1) namespace: kube-system:
[2021-11-30T03:19:47.656Z]   Updating resource deployment/calico-kube-controllers (apps/v1) namespace: kube-system:
[2021-11-30T03:19:47.656Z]     Deployment.apps "calico-kube-controllers" is invalid: spec.selector:
[2021-11-30T03:19:47.656Z]       Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"k8s-app":"calico-kube-controllers"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable (reason: Invalid), retrying
```
From the log you can see that it is trying to update a "clean" LabelSelector, which do not contain kapp LabelSelectors. This error showed up because the previous version of calico that installed in workload cluster do not have the `disable-default-label-scoping-rules`, that means it contains kapp LabelSelector, and after we add `disable-default-label-scoping-rules` in new version of calico, during update it will try to remove those kapp LabelSelector, which is immutable.

Then we added this `update-strategy` annotation to let it replace the deployment of calico-kube-controllers who has kapp LabelSelector, and the problem is fixed. If we do not add this, the upgrade of workload cluster becomes a breaking change which will always fail.

Comment (vijaykatam) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: I see, in that case can we change update-strategy to fallback-on-replace?

Comment (12345lcr) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: https://github.com/projectcalico/calico/blob/master/getting-started/kubernetes/installation/calico-kube-controllers.yaml#L19-L25
Hey @vijaykatam , I found in calico manifest that calico-kube-controllers deployment already set the strategy to `Recreate` due to only one single instance of the this pod should be active at a time. Otherwise there will be potential problems like reconciling at the same time and creating duplicate resources.

This means during the update, the pod will be deleted first, then a new one is recreated. There is no `RollingUpdate` situation that we need to take care of.

So from the pod level, basically there is no difference between `fallback-on-replace` and `always-replace`, the pod will always be deleted and recreated no matter which update-strategy we choose. Since the `always-replace` strategy is fully tested and we are approaching release date, I suggest we keep this as it is to follow the `Recreate` pattern. And if anything unexpected happened let's create another issue to track, triage and fix. What do you think?

Comment (vijaykatam) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: For pod level there is no difference but the deployment could have other objects associated to it, in which case there is difference between kapp delete deployment object on `always-replace` and deployment deleting pod object on `Recreate`. You are going for nuclear options always but looks like `calico-kube-controllers` doesn't have any other objects such as configmaps or secrets associated to it so in this case it is identical. 

It looks like you did a bunch of testing and want to get this merged and I am not opposed to it but in the future have me review upstream TCE PR so that we can have it resolved there before you do a bunch of downstream integration testing.

**Review by vijaykatam — APPROVED** [review]

File: pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml
Comment (12345lcr) [reply pkg/v1/providers/ytt/vendir/cni/_ytt_lib/addons/packages/calico/3.19.1/bundle/config/overlay/calico_overlay.yaml:17]: Totally agree. Currently in tce there is also a concept to standardize the update strategy.
https://github.com/vmware-tanzu/community-edition/pull/2277

**Pull request merged**
